### DISCUSSION TITLE
Generating code from DB and making that a build step.

### DISCUSSION THREAD

**Discussion by zetashift** [aws/aws-pdk#378] [Q&A] [answered]
I have a `nx-monorepo` using `type-safe-api` and a `CdkApp` and I'm using https://www.edgedb.com/. This database has 2 commands, one to generate a (type-safe) query builder and the other to generate TS interfaces.

The commands look like this(run from the root of the `nx-monorepo`):

To generate the query builder
```
pnpx @edgedb/generate edgeql-js --output-dir ./packages/mycdkproject/edgeql-js 
```

To generate the interfaces 
```
pnpx @edgedb/generate queries --file ./packages/mycdkproject/data-access/interfaces.ts
```

What I would like is to add these 2 generate tasks, but also include them in the build step _before_ the `cdk-app` is compiled(because I use the query builder and interfaces in the Lambda).

Is the best way here to just create a new TypeScript project(where the build step is running those tasks) and make it a dependency of my `cdk-app`?

**Comment by zetashift** [comment]
Also another database related question, tho it's more of Lambda question.

I'm using `handlerRouter` and I was wondering what the best way was to handle the DB client. The server handles connection pooling, but currently I have a static method that instantiates a new db client:

```typescript
export class DBClient {
  /** Creates a new connection to our EdgeDB server for data access
   *
   * @returns a new client connected to our EdgeDB instance
   */
  public static create(connString: string | undefined = undefined): edgedb.Client {
    const client = edgedb.createClient(connString);
    return client;
  }
}
```

But I call `DBClient.create` in several files like `foo.ts`, which defines all the handlers for `Foo` and `bar.ts` , which defines all the handlers for `Bar`. The handlers then get aggregated by the `handlerRouter` and put in a single lambda. I can imagine this might cause performance issues down the line because several clients could be created?

**Reply by cogwirrel** [reply]
Yeah it looks like if you used `DBClient.create` in multiple places there might be multiple clients hanging around (if the same container is spun up to serve multiple requests). Usually if you only want a single shared client you can use the singleton pattern, eg:

```ts
export class DBClient {
  private static client: edgedb.Client | undefined;

  public static get(): edgedb.Client {
    if (!DBClient.client) {
      DBClient.client = edgedb.createClient();
    }
    return DBClient.client;
  }
}
```

If your connection string changes you may need to consider keeping a singleton client per connection string, in eg a `private static clients: { [connString: string]: edgedb.Client };`

Hope that helps! :)

**Reply by zetashift** [reply]
Ah! That is exactly what I was looking for, never thought about just simply making it an if statement lol!

Thank you!

**Comment by cogwirrel** [comment, marked as answer]
> Is the best way here to just create a new TypeScript project(where the build step is running those tasks) and make it a dependency of my cdk-app?

I think that's the best way! If you generate your code into a `TypeScriptProject` managed by projen and part of the monorepo, it makes it fairly straightforward to add dependencies! Your suggestion is basically how TypeSafeApi works under the hood - where code generation is run as part of the `preCompile` task for the relevant projects :)

**Reply by zetashift** [reply]
Sounds good, I guess this is one of the nice things of the monorepo :D
